lib/webpacker/file_loader.rb

```diff
@@ -4,21 +4,38 @@ class NotFoundError < StandardError; end
   class FileLoaderError < StandardError; end
 
   class_attribute :instance
-  attr_accessor :data
+  attr_accessor :data, :mtime, :path
 
   class << self
-    def load(path = file_path)
-      self.instance = new(path)
+    def load_instance(path = file_path)
+      if instance.nil? || !production_env? || !file_cached?(path)
+        self.instance = new(path)
+      end
     end
+
+    def file_path
+      raise FileLoaderError.new("Subclass of Webpacker::FileLoader should override this method")
+    end
+
+    private
+      def file_cached?(path)
+        File.exist?(path) && self.instance.mtime == File.mtime(path)
+      end
+
+      # Prefer the NODE_ENV to the rails env.
+      def production_env?
+        (ENV["NODE_ENV"].presence || Rails.env) == "production"
+      end
   end
 
   private
     def initialize(path)
       @path = path
-      @data = load
+      @mtime = File.exist?(path) ? File.mtime(path) : nil
+      @data = load_data
     end
 
-    def load
+    def load_data
```

Comment:
I find it very confusing to have `load` mean both load_instance and load_data, depending on if you have the static or non-static method. @dhh @gauravtiwari